components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/provisioning/impl/DefaultProvisioningHandler.java

```diff
@@ -718,7 +719,42 @@ private String getUserStoreDomain(String userStoreDomain, UserRealm realm)
      * @return
      */
     protected String generatePassword() {
-        return RandomStringUtils.randomNumeric(12);
+
+        // Pick from some letters that won't be easily mistaken for each other.
+        // So, for example, omit o O and 0, 1 l and L.
+        // This will generate a random password which satisfy the following regex.
+        // ^((?=.*\\d)(?=.*[a-z])(?=.*[A-Z])(?=.*[!@#$%&*])).{12}$}
+        Random secureRandom = new SecureRandom();
+        String characters = "23456789abcdefghjkmnpqrstuvwxyzABCDEFGHJKMNPQRSTUVWXYZ!@#$%&*";
+        String digits = "23456789";
+        String lowercaseLetters = "abcdefghjkmnpqrstuvwxyz";
+        String uppercaseLetters = "ABCDEFGHJKMNPQRSTUVWXYZ";
+        String specialCharacters = "!@#$%&*";
+        int passwordLength = 12;
+        int mandatoryCharactersCount = 4;
+
+        StringBuilder pw = new StringBuilder();
+        int index;
+        for (int i = 0; i < passwordLength - mandatoryCharactersCount; i++) {
+            index = secureRandom.nextInt(characters.length());
+            pw.append(characters.charAt(index));
+        }
+
+        index = secureRandom.nextInt(digits.length());
+        pw.append(digits.charAt(index));
```

Comment:
Instead of repeatedly reassigning the index variable, you can directly add characters from each category as below:

``` 
pw.append(digits.charAt(secureRandom.nextInt(digits.length())));`
pw.append(lowercaseLetters.charAt(secureRandom.nextInt(lowercaseLetters.length())));
pw.append(uppercaseLetters.charAt(secureRandom.nextInt(uppercaseLetters.length())));
pw.append(specialCharacters.charAt(secureRandom.nextInt(specialCharacters.length())));
```

Reply:
Addressed in the latest commit https://github.com/wso2/carbon-identity-framework/pull/5590/commits/e058432bae954566873729e3974a87a814c4d675